imsim/photon_ops.py

```diff
@@ -135,12 +119,248 @@ def applyTo(self, photon_array, local_wcs=None, rng=None):
+    and optional parameters.
+    """
+    req, opt, single, _takes_rng = get_cls_params(cls)
+    kwargs, _safe = GetAllParams(config, base, req, opt, single)
+    opt_kwargs = specified_opt_kwargs(kwargs, cls._opt_params.keys())
+    return kwargs, opt_kwargs
+
+
+@photon_op_type("lsst_optics")
+def deserialize_lsst_optics(config, base, _logger):
+    kwargs, opt_kwargs = config_kwargs(config, base, LsstOptics)
+    shift_optics = opt_kwargs.pop("shift_optics", base.get("shift_optics", None))
+
+    return LsstOptics(
+        telescope=base["_telescope"],
+        boresight=kwargs["boresight"],
+        sky_pos=base["sky_pos"],
+        image_pos=base["image_pos"],
+        icrf_to_field=base["_icrf_to_field"],
+        det_name=base["det_name"],
+        shift_optics=shift_optics,
+        **opt_kwargs,
+    )
+
+
+@photon_op_type("lsst_diffraction")
+def deserialize_lsst_diffraction(config, base, _logger):
+    kwargs, opt_kwargs = config_kwargs(config, base, LsstDiffraction)
+    shift_optics = opt_kwargs.pop("shift_optics", base.get("shift_optics", None))
+
+    opsim_meta = galsim.config.GetInputObj(
+        "opsim_meta_dict", config, base, "opsim_meta_dict"
+    )
+
+    return LsstDiffraction(
+        telescope=base["_telescope"],
+        latitude=kwargs["latitude"],
+        altitude=opsim_meta["altitude"],
+        azimuth=opsim_meta["azimuth"],
+        sky_pos=base["sky_pos"],
+        icrf_to_field=base["_icrf_to_field"],
+        shift_optics=shift_optics,
+        **opt_kwargs,
+    )
+
+
+class XyToV:
+    """Maps image coordinates (x,y) to a 3d cartesian direction
+    vector of the photons before entering the telescope.
+
+    The transform is composed of the chain
+    (x,y) -> (u,v) -> (ra,dec) -> (thx, thy) -> v.
+
+    The transform takes 2 vectors of shape (n,) and returns
+    a column major vector of shape (n,3).
+    """
+
+    def __init__(self, local_wcs, icrf_to_field, sky_pos):
+        self.local_wcs = local_wcs
+        self.icrf_to_field = icrf_to_field
+        self.sky_pos = sky_pos
+
+    def __call__(self, x: np.ndarray, y: np.ndarray) -> np.ndarray:
+        # x/y to u/v
+        u = self.local_wcs._u(x, y)
+        v = self.local_wcs._v(x, y)
+        # u/v to ICRF
+        u_rad = np.deg2rad(u / 3600)
+        v_rad = np.deg2rad(v / 3600)
+        ra, dec = self.sky_pos.deproject_rad(u_rad, v_rad)
+        # ICRF to field
+        thx, thy = self.icrf_to_field.radecToxy(ra, dec, units="rad")
+
+        return np.array(batoid.utils.gnomonicToDirCos(thx, thy)).T
+
+    def inverse(self, v: np.ndarray) -> tuple:
+        thx, thy = batoid.utils.dirCosToGnomonic(v[:, 0], v[:, 1], v[:, 2])
+        # field to ICRF
+        ra, dec = self.icrf_to_field.xyToradec(thx, thy, units="rad")
+        # ICRF to u/v
+        u_rad, v_rad = self.sky_pos.project_rad(ra, dec)
+        u = 3600 * np.rad2deg(u_rad)
+        v = 3600 * np.rad2deg(v_rad)
```

Comment:
You're reusing the name `v` here.  That's not a good idea.  Can we come up a different name for the 3d velocity vector?  u,v are used all over the place to mean the tangent plane position, so I think it's better to leave this v alone and change the other one.  Maybe `dir` for direction, since I believe it's a unit vector representing the direction the photon is moving, right?

Reply:
`dir` is a builtin function. would `v_photon` or `v_phot` be ok for you?

Reply:
Anything to disambiguate is fine with me.